src/dmd/escape.d

```diff
@@ -154,71 +154,52 @@ bool checkMutableArguments(Scope* sc, FuncDeclaration fd, TypeFunction tf,
             escapeByValue(arg, &eb.er);
     }
 
-    foreach (const i, ref eb; escapeBy[0 .. $ - 1])
+    void checkOnePair(size_t i,ref EscapeBy eb, ref EscapeBy eb2,
+                      VarDeclaration v, VarDeclaration v2, bool of)
     {
-        foreach (VarDeclaration v; eb.er.byvalue)
+        if (log) printf("v2: `%s`\n", v2.toChars());
+        if (v2 != v)
+            return;
+        //printf("v %d v2 %d\n", eb.isMutable, eb2.isMutable);
+        if (!(eb.isMutable || eb2.isMutable))
+            return;
+        if (!(global.params.vsafe && sc.func.setUnsafe()))
+            return;
+
+        if (!gag)
         {
-            if (log) printf("byvalue `%s`\n", v.toChars());
-            if (!v.type.hasPointers())
-                continue;
-            foreach (ref eb2; escapeBy[i + 1 .. $])
-            {
-                foreach (VarDeclaration v2; eb2.er.byvalue)
-                {
-                    if (log) printf("v2: `%s`\n", v2.toChars());
-                    if (v2 != v)
-                        continue;
-                    if (eb.isMutable || eb2.isMutable)
-                    {
-                        if (global.params.vsafe && sc.func.setUnsafe())
-                        {
-                            if (!gag)
-                            {
-                                const(char)* msg = eb.isMutable && eb2.isMutable
-                                    ? "more than one mutable reference of `%s` in arguments to `%s()`"
-                                    : "mutable and const references of `%s` in arguments to `%s()`";
-                                error((*arguments)[i].loc, msg,
-                                    v.toChars(),
-                                    fd ? fd.toPrettyChars() : "indirectly");
-                            }
-                            errors = true;
-                        }
-                    }
-                }
-            }
+            const(char)* msg = eb.isMutable && eb2.isMutable
+            ? "more than one mutable reference %s `%s` in arguments to `%s()`"
+            : "mutable and const references %s `%s` in arguments to `%s()`";
```

Comment:
(ignoring the fact that this doesn't compile as is) I'm wondering why we give two different, but almost identical, error messages here. I for one am all for giving the same message.

Reply:
What message would that be ?

Reply:
byval escaping ref print `reference to`. escaping  by ref prints `reference of`. It took me a while to see what the difference was when looking at the error message.

Reply:
> I'm wondering why we give two different, but almost identical, error messages here.

I often make them slightly different as it makes it easier to find which one is tripping.

Reply:
Fair enough.